src/R/Components/Impl/PackageManager/Implementation/ViewModel/RPackageManagerViewModel.cs

```diff
@@ -321,25 +303,19 @@ internal class RPackageManagerViewModel : BindableBase, IRPackageManagerViewMode
             }
 
             await ReloadLoadedPackagesAsync();
-
-            AfterLoadUnload(package);
+            await AfterLoadUnloadAsync(package);
         }
 
-        private void BeforeLoadUnload() {
+        private void BeforeLoadUnload(IRPackageViewModel package) {
             if (_selectedTab == SelectedTab.InstalledPackages || _selectedTab == SelectedTab.LoadedPackages) {
                 IsLoading = true;
             }
+            package.IsChanging = true;
         }
 
-        private void AfterLoadUnload(IRPackageViewModel package) {
+        private async Task AfterLoadUnloadAsync(IRPackageViewModel package) {
+            await ReplaceItemsAsync();
```

Comment:
Here is the test (add it to `RPackageManagerViewModelTest`):

```{csharp}
        [Test(ThreadType.UI)]
        public async Task SwitchFromInstalledToAvailableWhileLoadingInstalled() {
            // We need real repo for this test
            await TestRepositories.SetCranRepoAsync(_workflow.RSession);

            _packageManagerViewModel.SwitchToLoadedPackagesAsync().DoNotWait();
            var t1 = _packageManagerViewModel.SwitchToInstalledPackagesAsync();
            var t2 = _packageManagerViewModel.SwitchToAvailablePackagesAsync();

            await t1;
            _packageManagerViewModel.IsLoading.Should().BeTrue();

            await t2;
            _packageManagerViewModel.IsLoading.Should().BeFalse();
        }
```

```{csharp}
        public static Task SetCranRepoAsync(IRExpressionEvaluator eval) {
            var code = $"options(repos=list(CRAN=\"[http://cran.r-project.org\](http://cran.r-project.org\)"))";
            return eval.ExecuteAsync(code);
        }
```

---
In reply to: [73560560](https://github.com/Microsoft/RTVS/pull/2149#discussion_r73560560) [](ancestors = 73560560,73554727)